instrumentation/rocketmq-client/src/main/java/brave/rocketmq/client/RocketMQTracing.java

```diff
@@ -0,0 +1,147 @@
+    String remoteServiceName) {
+    return new RocketMQTracing(messagingTracing, remoteServiceName);
+  }
+
+  final Tracing tracing;
+  final Tracer tracer;
+  final Extractor<TracingProducerRequest> producerExtractor;
+  final Extractor<TracingConsumerRequest> consumerExtractor;
+  final Injector<TracingProducerRequest> producerInjector;
+  final Injector<TracingConsumerRequest> consumerInjector;
+  final String[] traceIdHeaders;
+  final SamplerFunction<MessagingRequest> producerSampler, consumerSampler;
+  final String remoteServiceName;
+
+  RocketMQTracing(MessagingTracing messagingTracing,
+    String remoteServiceName) { // intentionally hidden constructor
+    this.tracing = messagingTracing.tracing();
+    this.tracer = tracing.tracer();
+    Propagation<String> propagation = messagingTracing.propagation();
+    this.producerExtractor = propagation.extractor(TracingProducerRequest.GETTER);
+    this.consumerExtractor = propagation.extractor(TracingConsumerRequest.GETTER);
+    this.producerInjector = propagation.injector(TracingProducerRequest.SETTER);
+    this.consumerInjector = propagation.injector(TracingConsumerRequest.SETTER);
+    this.producerSampler = messagingTracing.producerSampler();
+    this.consumerSampler = messagingTracing.consumerSampler();
+    this.remoteServiceName = remoteServiceName;
+
+    // We clear the trace ID headers, so that a stale consumer span is not preferred over current
+    // listener. We intentionally don't clear BaggagePropagation.allKeyNames as doing so will
+    // application fields "user_id" or "country_code"
+    this.traceIdHeaders = propagation.keys().toArray(new String[0]);
+  }
+
+  <R> TraceContextOrSamplingFlags extractAndClearTraceIdHeaders(Extractor<R> extractor,
+    R request,
+    Map<String, String> properties) {
+    TraceContextOrSamplingFlags extracted = extractor.extract(request);
+    // Clear any propagation keys present in the headers
+    if (extracted.samplingFlags() == null) { // then trace IDs were extracted
+      if (properties != null) {
+        clearTraceIdHeaders(properties);
+      }
+    }
+    return extracted;
+  }
+
+  /** Creates a potentially noop remote span representing this request. */
+  Span nextMessagingSpan(SamplerFunction<MessagingRequest> sampler, MessagingRequest request,
+    TraceContextOrSamplingFlags extracted) {
+    Boolean sampled = extracted.sampled();
+    // only recreate the context if the messaging sampler made a decision
+    if (sampled == null && (sampled = sampler.trySample(request)) != null) {
+      extracted = extracted.sampled(sampled);
+    }
+    return tracer.nextSpan(extracted);
+  }
+
+  // We can't just skip clearing headers we use because we might inject B3 single, yet have stale B3
+  // multi, or vice versa.
+  void clearTraceIdHeaders(Map<String, String> headers) {
+    for (String traceIDHeader : traceIdHeaders)
+      headers.remove(traceIDHeader);
+  }
+
+  public Tracing tracing() {
```

Comment:
we don't need to expose this, as someone can use the same `MessagingTracing` to get this property used to construct this object, to get `tracing`. so remove it please.